numpy/core/src/multiarray/_multiarray_tests.c.src

```diff
@@ -601,14 +601,20 @@ static PyObject *
 fromstring_null_term_c_api(PyObject *dummy, PyObject *byte_obj)
 {
     char *string;
+    PyObject *dtype, *ret;
 
     string = PyBytes_AsString(byte_obj);
     if (string == NULL) {
         return NULL;
     }
-
-    return PyArray_FromString(
-                    string, -1, PyArray_DescrFromType(NPY_FLOAT64), -1, " ");
+    dtype = PyArray_DescrFromType(NPY_FLOAT64);
+    Py_INCREF(dtype);
+    ret = PyArray_FromString( string, -1, dtype, -1, " ");
+    if (ret == NULL)
+    {
+        Py_DECREF(dtype);
+    }
+    return ret;
```

Comment:
this was one of the places that produced the error message, and was straightforward to find

Reply:
`PyArray_DescrFromType` looks weird, it returns a new reference for strings, returns a borrowed reference otherwise.

Reply:
That function basically knows that some dtypes are immutable, while flexible dtypes are sometimes "adapted" so they need to be copied...

Reply:
Wait, it returns a borrowed reference? I thought it always increfs, now I am confused :), have to look closer at it later...

Reply:
Ah, sorry. I keep confusing `PyArray_NewFromDescr` and this one (if you want to be clean, add the NULL check). Isn't it enough to replace this with NewFromDescr (which practically never fails, except for flexible types)? The cleanup should not be necessary, since FromString has to steal the reference.

Reply:
Sorry: `PyArray_DescrNewFromType`

Reply:
fixed